libs/cardano-ledger-test/src/Test/Cardano/Ledger/Generic/Proof.hs

```diff
@@ -363,3 +412,96 @@ instance Shaped Proof any where
   shape (Alonzo c) = Nary 3 [shape c]
   shape (Babbage c) = Nary 4 [shape c]
   shape (Conway c) = Nary 5 [shape c]
+
+-- ======================================================
+-- GADT's that witness the special properties that hold
+-- for some type families
+
+data TxOutWit era where
+  TxOutShelleyToMary :: (TxOut era ~ ShelleyTxOut era, EraTxOut era, ProtVerAtMost era 8) => TxOutWit era
+  TxOutAlonzoToAlonzo :: (TxOut era ~ AlonzoTxOut era, AlonzoEraTxOut era, ProtVerAtMost era 8) => TxOutWit era
+  TxOutBabbageToConway :: (TxOut era ~ BabbageTxOut era, BabbageEraTxOut era) => TxOutWit era
+
+whichTxOut :: Proof era -> TxOutWit era
+whichTxOut (Shelley _) = TxOutShelleyToMary
+whichTxOut (Allegra _) = TxOutShelleyToMary
+whichTxOut (Mary _) = TxOutShelleyToMary
+whichTxOut (Alonzo _) = TxOutAlonzoToAlonzo
+whichTxOut (Babbage _) = TxOutBabbageToConway
+whichTxOut (Conway _) = TxOutBabbageToConway
+
+data TxCertWit era where
+  TxCertShelleyToBabbage :: (TxCert era ~ ShelleyTxCert era, ShelleyEraTxCert era, ProtVerAtMost era 8) => TxCertWit era
+  TxCertConwayToConway :: (TxCert era ~ ConwayTxCert era, ConwayEraTxCert era, ProtVerAtLeast era 9) => TxCertWit era
```

Comment:
I am pretty sure you don't need this constraint
```suggestion
  TxCertConwayToConway :: (TxCert era ~ ConwayTxCert era, ConwayEraTxCert era) => TxCertWit era
```

Reply:
done